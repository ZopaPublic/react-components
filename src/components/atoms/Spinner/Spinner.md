### Summary

Use `<Spinner />` whenever we need to notify the user that something is loading.

### Examples

Default:

```ts
import { Spinner } from '@zopauk/react-components';

<Spinner />;
```

Small:

```ts
import { Spinner } from '@zopauk/react-components';

<Spinner size="small" />;
```

Negative:

```ts { "props": { "style": { "backgroundColor": "black", "border": "none" } } }
import { Spinner } from '@zopauk/react-components';

<Spinner styling="negative" />;
```
