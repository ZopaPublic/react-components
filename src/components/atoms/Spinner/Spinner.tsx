import React from 'react';
import deprecated from 'deprecated-prop-type';
import brandSpinner from './spinner-brand.gif';
import actionSpinner from './spinner-action.gif';
import negativeSpinner from './spinner-negative.gif';
import { useThemeContext } from '../../styles/Theme';
import CustomSpinner from './CustomSpinner/CustomSpinner';

export type SpinnerStyling = 'primary' | 'secondary' | 'negative';

export interface SpinnerProps {
  /**
   * Size of the spinner
   * @default 'standard'
   */
  size?: 'standard' | 'small';
  /**
   * Determines the styling of the spinner
   * @default 'primary'
   */
  styling?: SpinnerStyling;
  /**
   * Adjusts colour to display on top of dark background
   * @default false
   */
  negative?: boolean;
}

const animationData: Record<SpinnerStyling, string> = {
  primary: brandSpinner,
  secondary: actionSpinner,
  negative: negativeSpinner,
};

const Spinner = (props: SpinnerProps) => {
  const theme = useThemeContext();

  deprecated(props.negative, 'negative prop is deprecated, use styling="negative" instead');

  const spinner = props.negative ? negativeSpinner : animationData[props.styling || 'primary'];

  const size = props.size === 'small' ? '20px' : '40px';

  return theme.spinner.spinnerTheme !== 'zopa' && theme.spinner.customSpinner ? (
    <CustomSpinner size={props.size} styling={props.styling} />
  ) : (
    <img src={spinner} height={size} width={size} aria-label="loading spinner" />
  );
};

export default Spinner;
