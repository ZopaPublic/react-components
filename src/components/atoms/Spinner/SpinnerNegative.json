{
  "v": "5.5.2",
  "fr": 29.9700012207031,
  "ip": 0,
  "op": 187.000007616666,
  "w": 260,
  "h": 260,
  "nm": "spinner",
  "ddd": 0,
  "assets": [
    {
      "id": "comp_0",
      "layers": [
        {
          "ddd": 0,
          "ind": 1,
          "ty": 4,
          "nm": "spinner_04",
          "sr": 1,
          "ks": {
            "o": { "a": 0, "k": 100, "ix": 11 },
            "r": {
              "a": 1,
              "k": [
                { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 141, "s": [-90] },
                { "t": 189.000007698128, "s": [270] }
              ],
              "ix": 10
            },
            "p": { "a": 0, "k": [147, 156, 0], "ix": 2 },
            "a": { "a": 0, "k": [147, 155, 0], "ix": 1 },
            "s": { "a": 0, "k": [100, 100, 100], "ix": 6 }
          },
          "ao": 0,
          "shapes": [
            {
              "ty": "gr",
              "it": [
                {
                  "ind": 0,
                  "ty": "sh",
                  "ix": 1,
                  "ks": {
                    "a": 0,
                    "k": {
                      "i": [
                        [0, -49.982],
                        [49.981, 0],
                        [0, 49.981],
                        [-49.982, 0]
                      ],
                      "o": [
                        [0, 49.981],
                        [-49.982, 0],
                        [0, -49.982],
                        [49.981, 0]
                      ],
                      "v": [
                        [90.5, 0],
                        [0, 90.5],
                        [-90.5, 0],
                        [0, -90.5]
                      ],
                      "c": true
                    },
                    "ix": 2
                  },
                  "nm": "Path 1",
                  "mn": "ADBE Vector Shape - Group",
                  "hd": false
                },
                {
                  "ty": "st",
                  "c": { "a": 0, "k": [1, 1, 1, 1], "ix": 3 },
                  "o": { "a": 0, "k": 100, "ix": 4 },
                  "w": { "a": 0, "k": 32, "ix": 5 },
                  "lc": 2,
                  "lj": 1,
                  "ml": 10,
                  "bm": 0,
                  "nm": "Stroke 1",
                  "mn": "ADBE Vector Graphic - Stroke",
                  "hd": false
                },
                {
                  "ty": "tr",
                  "p": { "a": 0, "k": [147, 155], "ix": 2 },
                  "a": { "a": 0, "k": [0, 0], "ix": 1 },
                  "s": { "a": 0, "k": [100, 100], "ix": 3 },
                  "r": { "a": 0, "k": 0, "ix": 6 },
                  "o": { "a": 0, "k": 100, "ix": 7 },
                  "sk": { "a": 0, "k": 0, "ix": 4 },
                  "sa": { "a": 0, "k": 0, "ix": 5 },
                  "nm": "Transform"
                }
              ],
              "nm": "Group 1",
              "np": 2,
              "cix": 2,
              "bm": 0,
              "ix": 1,
              "mn": "ADBE Vector Group",
              "hd": false
            },
            {
              "ty": "tm",
              "s": {
                "a": 1,
                "k": [
                  { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 153, "s": [0] },
                  { "t": 189.000007698128, "s": [100] }
                ],
                "ix": 1
              },
              "e": {
                "a": 1,
                "k": [
                  { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 141, "s": [1] },
                  { "t": 171.000006964973, "s": [100] }
                ],
                "ix": 2
              },
              "o": { "a": 0, "k": 0, "ix": 3 },
              "m": 1,
              "ix": 2,
              "nm": "Trim Paths 1",
              "mn": "ADBE Vector Filter - Trim",
              "hd": false
            }
          ],
          "ip": 141.000005743048,
          "op": 190.000007738859,
          "st": 141.000005743048,
          "bm": 0
        },
        {
          "ddd": 0,
          "ind": 2,
          "ty": 4,
          "nm": "spinner_03",
          "sr": 1,
          "ks": {
            "o": { "a": 0, "k": 100, "ix": 11 },
            "r": {
              "a": 1,
              "k": [
                { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 94, "s": [-90] },
                { "t": 142.000005783779, "s": [270] }
              ],
              "ix": 10
            },
            "p": { "a": 0, "k": [147, 155, 0], "ix": 2 },
            "a": { "a": 0, "k": [147, 155, 0], "ix": 1 },
            "s": { "a": 0, "k": [100, 100, 100], "ix": 6 }
          },
          "ao": 0,
          "shapes": [
            {
              "ty": "gr",
              "it": [
                {
                  "ind": 0,
                  "ty": "sh",
                  "ix": 1,
                  "ks": {
                    "a": 0,
                    "k": {
                      "i": [
                        [0, -49.982],
                        [49.981, 0],
                        [0, 49.981],
                        [-49.982, 0]
                      ],
                      "o": [
                        [0, 49.981],
                        [-49.982, 0],
                        [0, -49.982],
                        [49.981, 0]
                      ],
                      "v": [
                        [90.5, 0],
                        [0, 90.5],
                        [-90.5, 0],
                        [0, -90.5]
                      ],
                      "c": true
                    },
                    "ix": 2
                  },
                  "nm": "Path 1",
                  "mn": "ADBE Vector Shape - Group",
                  "hd": false
                },
                {
                  "ty": "st",
                  "c": { "a": 0, "k": [1, 1, 1, 1], "ix": 3 },
                  "o": { "a": 0, "k": 100, "ix": 4 },
                  "w": { "a": 0, "k": 32, "ix": 5 },
                  "lc": 2,
                  "lj": 1,
                  "ml": 10,
                  "bm": 0,
                  "nm": "Stroke 1",
                  "mn": "ADBE Vector Graphic - Stroke",
                  "hd": false
                },
                {
                  "ty": "tr",
                  "p": { "a": 0, "k": [147, 155], "ix": 2 },
                  "a": { "a": 0, "k": [0, 0], "ix": 1 },
                  "s": { "a": 0, "k": [100, 100], "ix": 3 },
                  "r": { "a": 0, "k": 0, "ix": 6 },
                  "o": { "a": 0, "k": 100, "ix": 7 },
                  "sk": { "a": 0, "k": 0, "ix": 4 },
                  "sa": { "a": 0, "k": 0, "ix": 5 },
                  "nm": "Transform"
                }
              ],
              "nm": "Group 1",
              "np": 2,
              "cix": 2,
              "bm": 0,
              "ix": 1,
              "mn": "ADBE Vector Group",
              "hd": false
            },
            {
              "ty": "tm",
              "s": {
                "a": 1,
                "k": [
                  { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 106, "s": [0] },
                  { "t": 142.000005783779, "s": [100] }
                ],
                "ix": 1
              },
              "e": {
                "a": 1,
                "k": [
                  { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 94, "s": [1] },
                  { "t": 124.000005050624, "s": [100] }
                ],
                "ix": 2
              },
              "o": { "a": 0, "k": 0, "ix": 3 },
              "m": 1,
              "ix": 2,
              "nm": "Trim Paths 1",
              "mn": "ADBE Vector Filter - Trim",
              "hd": false
            }
          ],
          "ip": 94.0000038286985,
          "op": 143.000005824509,
          "st": 94.0000038286985,
          "bm": 0
        },
        {
          "ddd": 0,
          "ind": 3,
          "ty": 4,
          "nm": "spinner_02",
          "sr": 1,
          "ks": {
            "o": { "a": 0, "k": 100, "ix": 11 },
            "r": {
              "a": 1,
              "k": [
                { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 47, "s": [-90] },
                { "t": 95.0000038694293, "s": [270] }
              ],
              "ix": 10
            },
            "p": { "a": 0, "k": [147, 155, 0], "ix": 2 },
            "a": { "a": 0, "k": [147, 155, 0], "ix": 1 },
            "s": { "a": 0, "k": [100, 100, 100], "ix": 6 }
          },
          "ao": 0,
          "shapes": [
            {
              "ty": "gr",
              "it": [
                {
                  "ind": 0,
                  "ty": "sh",
                  "ix": 1,
                  "ks": {
                    "a": 0,
                    "k": {
                      "i": [
                        [0, -49.982],
                        [49.981, 0],
                        [0, 49.981],
                        [-49.982, 0]
                      ],
                      "o": [
                        [0, 49.981],
                        [-49.982, 0],
                        [0, -49.982],
                        [49.981, 0]
                      ],
                      "v": [
                        [90.5, 0],
                        [0, 90.5],
                        [-90.5, 0],
                        [0, -90.5]
                      ],
                      "c": true
                    },
                    "ix": 2
                  },
                  "nm": "Path 1",
                  "mn": "ADBE Vector Shape - Group",
                  "hd": false
                },
                {
                  "ty": "st",
                  "c": { "a": 0, "k": [1, 1, 1, 1], "ix": 3 },
                  "o": { "a": 0, "k": 100, "ix": 4 },
                  "w": { "a": 0, "k": 32, "ix": 5 },
                  "lc": 2,
                  "lj": 1,
                  "ml": 10,
                  "bm": 0,
                  "nm": "Stroke 1",
                  "mn": "ADBE Vector Graphic - Stroke",
                  "hd": false
                },
                {
                  "ty": "tr",
                  "p": { "a": 0, "k": [147, 155], "ix": 2 },
                  "a": { "a": 0, "k": [0, 0], "ix": 1 },
                  "s": { "a": 0, "k": [100, 100], "ix": 3 },
                  "r": { "a": 0, "k": 0, "ix": 6 },
                  "o": { "a": 0, "k": 100, "ix": 7 },
                  "sk": { "a": 0, "k": 0, "ix": 4 },
                  "sa": { "a": 0, "k": 0, "ix": 5 },
                  "nm": "Transform"
                }
              ],
              "nm": "Group 1",
              "np": 2,
              "cix": 2,
              "bm": 0,
              "ix": 1,
              "mn": "ADBE Vector Group",
              "hd": false
            },
            {
              "ty": "tm",
              "s": {
                "a": 1,
                "k": [
                  { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 59, "s": [0] },
                  { "t": 95.0000038694293, "s": [100] }
                ],
                "ix": 1
              },
              "e": {
                "a": 1,
                "k": [
                  { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 47, "s": [1] },
                  { "t": 77.0000031362743, "s": [100] }
                ],
                "ix": 2
              },
              "o": { "a": 0, "k": 0, "ix": 3 },
              "m": 1,
              "ix": 2,
              "nm": "Trim Paths 1",
              "mn": "ADBE Vector Filter - Trim",
              "hd": false
            }
          ],
          "ip": 47.0000019143492,
          "op": 96.0000039101601,
          "st": 47.0000019143492,
          "bm": 0
        },
        {
          "ddd": 0,
          "ind": 4,
          "ty": 4,
          "nm": "spinner_01",
          "sr": 1,
          "ks": {
            "o": { "a": 0, "k": 100, "ix": 11 },
            "r": {
              "a": 1,
              "k": [
                { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 0, "s": [-90] },
                { "t": 48.0000019550801, "s": [270] }
              ],
              "ix": 10
            },
            "p": { "a": 0, "k": [147, 155, 0], "ix": 2 },
            "a": { "a": 0, "k": [147, 155, 0], "ix": 1 },
            "s": { "a": 0, "k": [100, 100, 100], "ix": 6 }
          },
          "ao": 0,
          "shapes": [
            {
              "ty": "gr",
              "it": [
                {
                  "ind": 0,
                  "ty": "sh",
                  "ix": 1,
                  "ks": {
                    "a": 0,
                    "k": {
                      "i": [
                        [0, -49.982],
                        [49.981, 0],
                        [0, 49.981],
                        [-49.982, 0]
                      ],
                      "o": [
                        [0, 49.981],
                        [-49.982, 0],
                        [0, -49.982],
                        [49.981, 0]
                      ],
                      "v": [
                        [90.5, 0],
                        [0, 90.5],
                        [-90.5, 0],
                        [0, -90.5]
                      ],
                      "c": true
                    },
                    "ix": 2
                  },
                  "nm": "Path 1",
                  "mn": "ADBE Vector Shape - Group",
                  "hd": false
                },
                {
                  "ty": "st",
                  "c": { "a": 0, "k": [1, 1, 1, 1], "ix": 3 },
                  "o": { "a": 0, "k": 100, "ix": 4 },
                  "w": { "a": 0, "k": 32, "ix": 5 },
                  "lc": 2,
                  "lj": 1,
                  "ml": 10,
                  "bm": 0,
                  "nm": "Stroke 1",
                  "mn": "ADBE Vector Graphic - Stroke",
                  "hd": false
                },
                {
                  "ty": "tr",
                  "p": { "a": 0, "k": [147, 155], "ix": 2 },
                  "a": { "a": 0, "k": [0, 0], "ix": 1 },
                  "s": { "a": 0, "k": [100, 100], "ix": 3 },
                  "r": { "a": 0, "k": 0, "ix": 6 },
                  "o": { "a": 0, "k": 100, "ix": 7 },
                  "sk": { "a": 0, "k": 0, "ix": 4 },
                  "sa": { "a": 0, "k": 0, "ix": 5 },
                  "nm": "Transform"
                }
              ],
              "nm": "Group 1",
              "np": 2,
              "cix": 2,
              "bm": 0,
              "ix": 1,
              "mn": "ADBE Vector Group",
              "hd": false
            },
            {
              "ty": "tm",
              "s": {
                "a": 1,
                "k": [
                  { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 12, "s": [0] },
                  { "t": 48.0000019550801, "s": [100] }
                ],
                "ix": 1
              },
              "e": {
                "a": 1,
                "k": [
                  { "i": { "x": [0.667], "y": [1] }, "o": { "x": [0.333], "y": [0] }, "t": 0, "s": [1] },
                  { "t": 30.0000012219251, "s": [100] }
                ],
                "ix": 2
              },
              "o": { "a": 0, "k": 0, "ix": 3 },
              "m": 1,
              "ix": 2,
              "nm": "Trim Paths 1",
              "mn": "ADBE Vector Filter - Trim",
              "hd": false
            }
          ],
          "ip": 0,
          "op": 49.0000019958109,
          "st": 0,
          "bm": 0
        }
      ]
    }
  ],
  "layers": [
    {
      "ddd": 0,
      "ind": 1,
      "ty": 0,
      "nm": "spin_circle",
      "refId": "comp_0",
      "sr": 1,
      "ks": {
        "o": { "a": 0, "k": 100, "ix": 11 },
        "r": { "a": 0, "k": 0, "ix": 10 },
        "p": { "a": 0, "k": [133, 127, 0], "ix": 2 },
        "a": { "a": 0, "k": [150, 150, 0], "ix": 1 },
        "s": { "a": 0, "k": [100, 100, 100], "ix": 6 }
      },
      "ao": 0,
      "w": 300,
      "h": 300,
      "ip": 0,
      "op": 188.000007657397,
      "st": 0,
      "bm": 0
    }
  ],
  "markers": []
}
