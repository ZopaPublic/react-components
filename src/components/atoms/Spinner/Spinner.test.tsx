import React from 'react';
import axe from '../../../../axe-helper';
import { render } from '@testing-library/react';
import Spinner from './Spinner';

describe('<Spinner />', () => {
  it('renders a spinner', async () => {
    const { container } = render(<Spinner />);
    expect(container.firstChild).toMatchSnapshot();
    const results = await axe(container.innerHTML);
    expect(results).toHaveNoViolations();
  });

  it('renders small', () => {
    const { container } = render(<Spinner size="small" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders negative', () => {
    const { container } = render(<Spinner negative />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
