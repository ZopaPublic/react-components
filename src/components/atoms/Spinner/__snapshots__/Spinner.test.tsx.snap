// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> renders a spinner 1`] = `
<img
  aria-label="loading spinner"
  height="40px"
  src="spinner-brand.gif"
  width="40px"
/>
`;

exports[`<Spinner /> renders negative 1`] = `
<img
  aria-label="loading spinner"
  height="40px"
  src="spinner-negative.gif"
  width="40px"
/>
`;

exports[`<Spinner /> renders small 1`] = `
<img
  aria-label="loading spinner"
  height="20px"
  src="spinner-brand.gif"
  width="20px"
/>
`;
